File: src/vnm/VirtualNetwork.cc
// ... existing code ...

#include <sstream>

// ... other includes ...

void VirtualNetwork::check_ovs_vlan_conflict()
{
    int tag;
    vector<int> trunks;
    bool has_tag = get_template_attribute("VLAN_ID", tag);
    bool has_trunks = get_template_attribute("VLAN_TRUNKS", trunks);

    if (has_tag && has_trunks)
    {
        for (size_t i = 0; i < trunks.size(); ++i)
        {
            if (trunks[i] == tag)
            {
                ostringstream oss;
                oss << "Warning: VLAN tag " << tag << " is also listed in VLAN_TRUNKS. This configuration is ambiguous in Open vSwitch.";
                NebulaLog::log("VNM", Log::WARNING, oss.str());
                break;
            }
        }
    }
}

// In the method that processes template update/create (e.g., add_template or post_update_template)
// Add call to check_ovs_vlan_conflict() after setting VLAN attributes.
// Example:

int VirtualNetwork::post_update_template()
{
    // ... existing logic ...
    check_ovs_vlan_conflict();
    // ... continue ...
}

// Similarly for create, call check_ovs_vlan_conflict() after parsing template.

File: src/oca/ruby/core/VirtualNetwork.rb
# VirtualNetwork.rb - Core VirtualNetwork class with OVS VLAN conflict check

class VirtualNetwork
  # ... existing code ...

  # Create a new virtual network
  def create(template)
    # ... existing create logic ...
    check_ovs_vlan_conflict(template)
    # ... rest of create ...
  end

  # Update an existing virtual network
  def update(template)
    # ... existing update logic ...
    check_ovs_vlan_conflict(template)
    # ... rest of update ...
  end

  private

  # Check for conflicting OVS VLAN tag and trunks configuration
  # Logs a warning if the VLAN tag is also listed in the trunks
  def check_ovs_vlan_conflict(template)
    driver = template['VNM_MAD'] || template[:vnm_mad]
    return unless driver && driver.downcase == 'ovswitch'

    vlan_id = template['VLAN_ID'] || template[:vlan_id]
    trunks_raw = template['VLAN_TRUNKS'] || template[:vlan_trunks]
    return unless vlan_id && trunks_raw

    # Parse trunks: assume comma-separated list of integers
    trunks = trunks_raw.to_s.split(',').map(&:strip).select { |t| t =~ /^\d+$/ }.map(&:to_i)
    vlan = vlan_id.to_i

    if trunks.include?(vlan)
      warn "[Warning] Virtual network VLAN tag #{vlan} is also configured as a trunk. This configuration is ambiguous in Open vSwitch. Consider removing #{vlan} from the trunks list."
    end
  end
end
